code/PaymentGatewayController.php

```diff
@@ -88,40 +99,149 @@ public function index()
+
+        if (!$payment) {
+            return $this->httpError(404, _t('Payment.NotFound', 'Payment could not be found.'));
+        }
+
+        $intent = $this->getPaymentIntent($payment->Status);
+
+        if (!$intent) {
+            return $this->httpError(403, _t('Payment.InvalidStatus', 'Invalid/unhandled payment status'));
+        }
+
+        $service = ServiceFactory::create()->getService($payment, $intent);
+        $service_response = $this->getServiceResponse($service, $this->request->param('Status'));
+
+        if (!$service_response) {
+            return $this->httpError(404, _t('Payment.InvalidUrl', 'Invalid payment url.'));
+        }
+        
+        $response = $service_response->redirectOrRespond();
+        
+        return $response;
+    }
+
+    /**
+     * Action used for handling static gateway requests
+     * (use this if your payment gateway doesnt handle
+     * dynamic callback URLs) 
+     */
+    public function gateway()
+    {
+        $response = null;
+        $gateway = $this->request->param("Gateway");
+        $identifier = null;
+
+        // Does the selected gateway allow static routes
+        if (!GatewayInfo::getConfigSetting($gateway, 'use_static_route')) {
+            return $this->httpError(404, _t('Payment.InvalidUrl', 'Invalid payment url.'));
+        }
+
+        $identifier = $this->getIdentifierFromRequest($this->request,$gateway);
+
+        if (!$identifier) {
+            return $this->httpError(404, _t('Payment.NotFound', 'Payment could not be found.'));
+        }
+
+        $payment = $this->getPaymentFromIdentifier($identifier);
+
+        if (!$payment) {
+            return $this->httpError(404, _t('Payment.NotFound', 'Payment could not be found.'));
+        }
+
+        $intent = $this->getPaymentIntent($payment->Status);
+
+        if (!$intent) {
+            return $this->httpError(403, _t('Payment.InvalidStatus', 'Invalid/unhandled payment status'));
+        }
+
+        $service = ServiceFactory::create()->getService($payment, $intent);
+        $service_response = $this->getServiceResponse($service, $this->request->param('Status'));
+
+        if (!$service_response) {
+            return $this->httpError(404, _t('Payment.InvalidUrl', 'Invalid payment url.'));
         }
 
+        $response = $service_response->redirectOrRespond();
+
         return $response;
     }
 
+    /**
+     * Attempt to get the the payment according to the identifier
+     * provided by the payment gateway.
+     * 
+     * Due to the many possible ways this can be retrieved, it is
+     * up to the implementer to extend this function and write their
+     *  
+     * 
+     * @param SS_HTTPRequest $request A SilverStripe request object
+     * @param string $gateway The identifier of the payment
+     * @return \Payment the payment
+     */
+    private function getIdentifierFromRequest(\SS_HTTPRequest $request, $gateway)
```

Comment:
Maybe a bit weird to look up the identifier, since you'll never know the identifier. You'd always have to look up the payment from the request data, and then return the payment-identifier from there… which in turn will be used to look up the payment again.
I'd probably stick with a method to get the payment from the request…